BenchmarkTests/Benchmarks/Sources/Metrics.swift

```diff
@@ -12,21 +12,133 @@ import QuartzCore
 let TASK_VM_INFO_COUNT = mach_msg_type_number_t(MemoryLayout<task_vm_info_data_t>.size / MemoryLayout<integer_t>.size)
 let TASK_VM_INFO_REV1_COUNT = mach_msg_type_number_t(MemoryLayout.offset(of: \task_vm_info_data_t.min_address)! / MemoryLayout<integer_t>.size)
 
-enum MachError: Error {
+internal enum MachError: Error {
     case task_info(return: kern_return_t)
     case task_threads(return: kern_return_t)
     case thread_info(return: kern_return_t)
 }
 
-public enum Memory {
+/// Aggregate metric values and compute `min`, `max`, `sum`, `avg`, and `count`.
+internal class MetricAggregator<T> where T: Numeric {
+    internal struct Aggregation {
+        let min: T
+        let max: T
+        let sum: T
+        let count: Int
+        let avg: Double
+    }
+
+    private var mutex = pthread_mutex_t()
+    private var _aggregation: Aggregation?
+
+    var aggregation: Aggregation? {
+        pthread_mutex_lock(&mutex)
+        defer { pthread_mutex_unlock(&mutex) }
+        return _aggregation
+    }
+
+    /// Resets the minimum frame rate to `nil`.
+    func reset() {
+        pthread_mutex_lock(&mutex)
+        _aggregation = nil
+        pthread_mutex_unlock(&mutex)
+    }
+
+    deinit {
+        pthread_mutex_destroy(&mutex)
+    }
+}
+
+extension MetricAggregator where T: BinaryInteger {
+    /// Records a `BinaryInteger` value.
+    ///
+    /// - Parameter value: The value to record.
+    func record(value: T) {
+        pthread_mutex_lock(&mutex)
+        _aggregation = _aggregation.map {
+            let sum = $0.sum + value
+            let count = $0.count + 1
+            return Aggregation(
+                min: Swift.min($0.min, value),
+                max: Swift.max($0.max, value),
+                sum: sum,
+                count: count,
+                avg: Double(sum) / Double(count)
+            )
+        } ?? Aggregation(min: value, max: value, sum: value, count: 1, avg: Double(value))
+        pthread_mutex_unlock(&mutex)
+    }
+}
+
+extension MetricAggregator where T: BinaryFloatingPoint {
```

Comment:
I was wondering if we could merge these two extensions since the logic inside is the same, but it doesn't seem straightforward.

Reply:
I tried, but failed :/ Problem is that there is 2 distinct `Double.init` that take either a `BinaryInteger` or a `BinaryFloatingPoint`. There is no common method that can help computing the `avg` value.